README.rst
==============
sc-autoencoder
==============

.. image:: https://github.com/mlf-core/sc-autoencoder/workflows/Train%20sc-autoencoder%20using%20CPU/badge.svg
        :target: https://github.com/mlf-core/sc-autoencoder/actions?query=workflow%3A%22Train+sc-autoencoder+using+CPU%22
        :alt: Github Workflow CPU Training sc-autoencoder Status

.. image:: https://github.com/mlf-core/sc-autoencoder/workflows/Publish%20Container%20to%20Docker%20Packages/badge.svg
        :target: https://github.com/mlf-core/sc-autoencoder/actions?query=workflow%3A%22Publish+Container+to+Docker+Packages%22
        :alt: Publish Container to Docker Packages

.. image:: https://github.com/mlf-core/sc-autoencoder/workflows/mlf-core%20linting/badge.svg
        :target: https://github.com/mlf-core/sc-autoencoder/actions?query=workflow%3A%22mlf-core+lint%22
        :alt: mlf-core lint


.. image:: https://readthedocs.org/projects/sc-autoencoder/badge/?version=latest
        :target: https://sc-autoencoder.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

An autoencoder for single cell data.

* Free software: MIT
* Documentation: https://sc-autoencoder.readthedocs.io.

This project uses an autoencoder model to learn latent features from single-cell RNA-seq (scRNA-seq) data. Autoencoder models
and similar architectures are frequently used for scRNA-seq data. For instance, Eraslan et al. used an `autoencoder <https://www.nature.com/articles/s41467-018-07931-2>`_ for denoising
of single cell data. In another study, Lotfollahi et al use a `variational autoencoder <https://www.nature.com/articles/s41592-019-0494-8>`_ to predict perturbation responses.
Here, we have implemented a very simple autoencoder to demonstrate how non-deterministic operations can lead to significant differences in latent space embeddings which affect
downstream analysis and hinder reproducibility.

Architecture
------------
The model used in this project follows a standard encoder-encoding-decoder autoencoder architecture.
We use layer sizes of 256, 128 and 64 for the encoder and decoder (in reverse) layers, and a encoding size of 32.


.. image:: docs/images/autoencoder_architecture.png
        :alt: Autoencoder architecture
        :scale: 10


Credits
-------

This package was created with `mlf-core`_ using Cookiecutter_.

.. _mlf-core: https://mlf-core.readthedocs.io/en/latest/
.. _Cookiecutter: https://github.com/audreyr/cookiecutter


CHANGELOG.rst
==========
Changelog
==========

This project adheres to `Semantic Versioning <https://semver.org/>`_.


1.0.0 (2021-03-08)
------------------

**Added**

* Created visualization folder containing plots and code to reproduce them
* Switched to newest mlf-core template [1.9.1]
* Improved documentation

**Fixed**

**Dependencies**

**Deprecated**


0.1.0-SNAPSHOT (2020-12-10)
---------------------------------------------

**Added**

* Created the project using mlf-core

**Fixed**

**Dependencies**

**Deprecated**


searchindex.js
Search.setIndex({docnames:["authors","changelog","code_of_conduct","index","model","readme","usage"],envversion:{"sphinx.domains.c":1,"sphinx.domains.changeset":1,"sphinx.domains.citation":1,"sphinx.domains.cpp":1,"sphinx.domains.index":1,"sphinx.domains.javascript":1,"sphinx.domains.math":2,"sphinx.domains.python":1,"sphinx.domains.rst":1,"sphinx.domains.std":1,"sphinx.ext.viewcode":1,sphinx:56},filenames:["authors.rst","changelog.rst","code_of_conduct.rst","index.rst","model.rst","readme.rst","usage.rst"],objects:{},objnames:{},objtypes:{},terms:{"default":4,"final":4,"float":6,"function":4,"int":6,"new":6,"public":2,"switch":1,"true":6,Added:1,Being:2,For:[4,5],The:[2,4,5,6],There:6,Using:2,abus:2,accept:2,access:6,account:2,act:2,action:2,activ:4,adam:4,adapt:2,address:2,adher:1,advanc:2,affect:5,age:2,align:2,all:[2,6],alon:6,analysi:5,ani:2,anoth:[5,6],appear:2,appli:2,appoint:2,appropri:2,architectur:3,attack:2,attent:2,attribut:3,automat:6,avail:[2,6],ban:2,base:[4,6],batch:6,behavior:2,benchmark:4,besid:6,best:2,bodi:2,both:2,build:6,can:[4,5,6],cannot:6,cell:5,changelog:3,chosen:4,circumst:2,clarifi:2,classic:4,classifi:4,code:[1,3],comment:2,commit:2,commun:2,complaint:2,conduct:3,confidenti:2,configur:6,connect:4,consid:2,construct:2,contain:[1,6],content:3,contribut:2,contributor:3,convolut:4,cookiecutt:5,core:[1,5,6],correct:2,could:2,coven:3,creat:[1,2,5,6],credit:3,critic:2,cuda:6,data:[3,5],dataset:4,decod:5,deem:2,defin:[2,6],demonstr:5,denois:5,depend:1,deprec:1,derogatori:2,desir:6,detail:[2,4],determin:2,determinist:5,develop:[3,6],differ:[2,5],digit:4,directori:4,disabl:2,document:[1,5,6],downstream:5,dropout:4,edit:2,either:6,electron:2,embed:5,empathi:2,encod:5,enforc:3,ensur:6,environ:[2,6],epoch:6,eraslan:5,ethnic:2,evalu:3,event:2,everi:6,everyon:2,exampl:2,expect:2,experi:2,explicit:2,express:2,face:2,fair:2,faith:2,featur:5,fetch:6,file:6,first:0,fix:1,flatten:4,focus:2,folder:1,follow:[2,4,5],foster:2,free:[2,5],frequent:5,from:[2,4,5],full:4,fulli:4,further:[2,4,6],gender:2,gener:[4,6],github:6,good:2,gracefulli:2,guarante:6,handwritten:4,harass:2,hardwar:6,harm:2,have:[2,5,6],here:[5,6],heumo:0,high:4,highli:6,hinder:5,histori:4,how:5,howev:6,html:2,http:[2,5],hyperparamet:3,ident:2,ids:6,imag:4,imageri:2,implement:5,improv:1,inappropri:2,incid:2,includ:2,inclus:2,index:3,individu:2,inform:2,insid:4,instal:6,instanc:[2,5],insult:2,intellig:6,interest:2,interfac:4,investig:2,issu:2,its:2,kevin:0,known:4,languag:2,latent:5,layer:[4,5],lead:[3,5],leadership:2,learn:[4,5,6],let:6,level:[2,4],linear:4,log:4,loss:4,lotfollahi:5,luka:0,mac:6,machin:4,mai:2,mail:2,maintain:2,make:2,manual:6,match:6,max:4,max_epoch:6,media:2,member:2,menden:0,mit:5,mlf:[1,5,6],mlflow:[4,6],mlproject:6,mnist:4,model:[3,5],modul:3,more:4,name:6,nation:2,need:6,net:0,newest:1,non:5,none:0,number:[4,6],numpi:6,nvidia:6,oblig:2,offens:2,offici:2,offlin:2,onlin:2,open:2,oper:5,optim:[4,6],org:2,orient:2,origin:4,other:2,otherwis:2,our:3,output:4,overview:3,packag:[5,6],page:3,part:4,particip:2,pass:6,perform:4,perman:2,permiss:2,person:2,perturb:5,physic:2,pledg:3,plot:1,polici:2,polit:2,pool:4,posit:2,post:2,posteo:0,predict:5,prefer:6,privat:2,profession:2,project:[1,2,4,5,6],publish:2,purpos:6,python:6,qbic:0,race:2,random:6,rate:6,readthedoc:5,reason:2,rectifi:4,regard:2,regardless:2,registri:6,reject:2,religion:2,remov:2,repercuss:2,replac:6,report:2,repres:2,represent:2,reproduc:[1,5],requir:6,respect:2,respond:2,respons:[3,5],revers:5,review:2,right:2,rna:5,root:4,run:[4,6],runtim:6,scope:3,scrna:5,search:3,seed:6,select:3,semant:1,separ:2,seq:5,set:[2,6],setup:3,sexual:2,should:6,show:2,signific:5,similar:5,simpl:5,sinc:6,singl:5,size:[2,5,6],snapshot:3,social:2,softmax:4,softwar:5,space:[2,5],specif:[2,6],standard:[3,5],start:6,string:6,strong:4,studi:5,suffici:6,summar:4,support:6,system:6,take:2,team:2,templat:1,temporari:2,temporarili:2,tensorflow:[4,6],test:[3,6],thei:2,them:1,thi:[1,2,4,5,6],threaten:2,todo:6,toolkit:6,toward:2,train:3,troll:2,tuebingen:0,unaccept:2,uni:0,unseen:4,unwelcom:2,usag:3,use:[2,5,6],used:[4,5],user:4,uses:5,using:[1,2,5],variat:5,veri:5,version:[1,2],via:2,viewabl:4,viewpoint:2,visit:4,visual:1,wai:2,want:6,websit:4,welcom:2,were:4,what:2,when:2,whether:6,which:[2,5],who:2,why:0,wide:4,wiki:2,window:6,within:2,without:2,write:6,written:4,www:2,yet:0,you:6,your:6},titles:["Credits","Changelog","Contributor Covenant Code of Conduct","Welcome to sc-autoencoder\u2019s documentation!","Model","sc-autoencoder","Usage"],titleterms:{architectur:[4,5],attribut:2,autoencod:[3,5],changelog:1,code:2,conda:6,conduct:2,contributor:[0,2],coven:2,cpu:6,credit:[0,5],data:4,develop:0,docker:6,document:3,enforc:2,evalu:4,gpu:6,hyperparamet:4,indic:3,lead:0,model:4,our:2,overview:4,paramet:6,pledg:2,respons:2,scope:2,select:4,setup:6,snapshot:1,standard:2,tabl:3,test:4,train:[4,6],usag:6,using:6,welcom:3}})

sc_autoencoder/sc_autoencoder.py
from argparse import ArgumentParser
import tensorflow as tf
import mlflow
import mlflow.tensorflow
import os
import time
from mlf_core.mlf_core import MLFCore
from rich import traceback
from data_loading.data_loader import load_data
from model.model import create_model
from training.train import train, test
from tensorflow.keras.mixed_precision import experimental as mp


def start_training():
    parser = ArgumentParser(description='Tensorflow example')
    parser.add_argument(
        '--cuda',
        type=bool,
        default=True,
        help='Enable or disable CUDA support',
    )
    parser.add_argument(
        '--max_epochs',
        type=int,
        default=10,
        help='Number of epochs to train',
    )
    parser.add_argument(
        '--general-seed',
        type=int,
        default=0,
        help='General Python, Python random and Numpy seed.',
    )
    parser.add_argument(
        '--tensorflow-seed',
        type=int,
        default=0,
        help='Tensorflow specific random seed.',
    )
    parser.add_argument(
        '--batch-size',
        type=int,
        default=64,
        help='Input batch size for training and testing',
    )
    parser.add_argument(
        '--buffer-size',
        type=int,
        default=10000,
        help='Buffer size for Mirrored Training',
    )
    parser.add_argument(
        '--lr',
        type=float,
        default=0.01,
        help='Learning rate',
    )
    parser.add_argument(
        '--mixed-precision',
        type=bool,
        default=False,
        help='Use mixed-prexicision training'
    )
    parser.add_argument(
        '--deterministic',
        type=str,
        default='1',
        help='Use GPU deterministic training'
    )
    args = parser.parse_args()
    dict_args = vars(args)

    # Disable GPU support if no GPUs are supposed to be used
    if not dict_args['cuda']:
        tf.config.set_visible_devices([], 'GPU')

    # Enable mixed precision training
    if dict_args['mixed_precision']:
        policy = mp.Policy('mixed_float16')
        mp.set_policy(policy)

    with mlflow.start_run():
        # Enable the logging of all parameters, metrics and models to mlflow and Tensorboard
        mlflow.autolog(1)

        # Log hardware and software
        MLFCore.log_sys_intel_conda_env()

        # Fix all random seeds and Tensorflow specific reproducibility settings
        MLFCore.set_general_random_seeds(dict_args["general_seed"])
        MLFCore.set_tensorflow_random_seeds(dict_args["tensorflow_seed"])
        MLFCore.set_tensorflow_random_seeds(dict_args["tensorflow_seed"], deterministic=dict_args['deterministic'])

        # Use Mirrored Strategy for multi GPU support
        strategy = tf.distribute.MirroredStrategy()
        print(f'[bold blue]Number of devices: {strategy.num_replicas_in_sync}')

        # Fetch and prepare dataset

        dataset, test_data = load_data(strategy, dict_args['batch_size'], dict_args['buffer_size'], dict_args['tensorflow_seed'])

        # Get the input dimension
        input_dim = 0
        for elem in dataset:
            input_dim = elem[0].shape[1]
            break

        with strategy.scope():

            # Define and compile model
            model = create_model(input_shape=input_dim)
            model.compile(loss=tf.keras.losses.MeanSquaredError(),
                          optimizer=tf.keras.optimizers.Adam(learning_rate=dict_args['lr']),
                          metrics=['mse'])

            model.build(input_shape=(dict_args['batch_size'], input_dim))
            # Train and evaluate the trained model
            runtime = time.time()

            train(model, dict_args['max_epochs'], dataset)
            embedding = test(model, test_data, save_path="embedding.png")
            mlflow.log_artifact(embedding + ".png")
            mlflow.log_artifact(embedding + ".csv")

            print(f'[bold blue]\nLaunch TensorBoard with:\ntensorboard --logdir={os.path.join(mlflow.get_artifact_uri(), "tensorboard_logs", "train")}')

            device = 'GPU' if dict_args['cuda'] else 'CPU'
            print(f'[bold green]{device} Run Time: {str(time.time() - runtime)} seconds')

            print(f'[bold blue]\nLaunch TensorBoard with:\ntensorboard --logdir={os.path.join(mlflow.get_artifact_uri(), "tensorboard_logs", "train")}')


if __name__ == '__main__':
    traceback.install()
    print(f'Num GPUs Available: {len(tf.config.experimental.list_physical_devices("GPU"))}')

    os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'  # Filtering out any Warnings messages

    start_training()


sc_autoencoder/training/train.py
import tensorflow as tf
import os
import umap
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np


def train(model, epochs, train_dataset):
    # Define the checkpoint directory to store the checkpoints
    checkpoint_dir = './training_checkpoints'
    checkpoint_prefix = os.path.join(checkpoint_dir, "ckpt_{epoch}")

    callbacks = [
        tf.keras.callbacks.TensorBoard(log_dir='./logs'),
        tf.keras.callbacks.ModelCheckpoint(filepath=checkpoint_prefix, save_weights_only=True)
    ]

    model.fit(train_dataset, epochs=epochs, callbacks=callbacks)
    model.load_weights(tf.train.latest_checkpoint(checkpoint_dir))


def test(model, eval_dataset, save_path="embedding"):

    # Get latent space encoding
    res = np.array(model.encode(eval_dataset))

    res_df = pd.DataFrame(res)
    res_df.to_csv(save_path + ".csv")

    # Reduce with UMAP
    reducer = umap.UMAP(random_state=42)
    embedding = reducer.fit_transform(res)
    df = pd.DataFrame(embedding, columns=['PC1', 'PC2'])
    plt.scatter(df['PC1'], df['PC2'])
    plt.savefig(save_path + ".png")

    return save_path


sc_autoencoder/mlf_core/mlf_core.py
import glob
import hashlib
import tempfile
import os
import numpy as np
import random
import subprocess
from rich import print
import mlflow
import tensorflow as tf


class MLFCore:
    __instance = None

    def __new__(cls):
        if cls.__instance is None:
            print('Creating the object')
            cls._instance = super(MLFCore, cls).__new__(cls)
        return cls._instance

    @staticmethod
    def set_general_random_seeds(seed):
        os.environ['PYTHONHASHSEED'] = str(seed)  # Python general
        np.random.seed(seed)  # Numpy random
        random.seed(seed)  # Python random

    @staticmethod
    def log_system_intelligence(reports_output_dir: str):
        # Scoped import to prevent issues like RuntimeError: Numba cannot operate on non-primary CUDA context
        from system_intelligence.query import query_and_export

        print(f'[bold blue]Writing reports locally to {reports_output_dir}\n')
        print('[bold blue]Running system-intelligence')
        query_and_export(query_scope={'all'},
                         verbose=False,
                         export_format='json',
                         generate_html_table=True,
                         output=f'{reports_output_dir}/system_intelligence.json')
        print('[bold blue]Uploading system-intelligence report as a run artifact...')
        mlflow.log_artifacts(reports_output_dir, artifact_path='reports')

    @staticmethod
    def log_conda_environment(reports_output_dir: str):
        print('[bold blue]Exporting conda environment...')
        conda_env_filehandler = open(f'{reports_output_dir}/sc_autoencoder_conda_environment.yml', 'w')
        subprocess.call(['conda', 'env', 'export', '--name', 'sc_autoencoder'], stdout=conda_env_filehandler)
        print('[bold blue]Uploading conda environment report as a run artifact...')
        mlflow.log_artifact(f'{reports_output_dir}/sc_autoencoder_conda_environment.yml', artifact_path='reports')

    @staticmethod
    def set_tensorflow_random_seeds(seed, deterministic='1'):
        tf.random.set_seed(seed)
        tf.config.threading.set_intra_op_parallelism_threads = 1  # CPU only
        tf.config.threading.set_inter_op_parallelism_threads = 1  # CPU only
        os.environ['TF_DETERMINISTIC_OPS'] = deterministic

    @classmethod
    def log_sys_intel_conda_env(cls):
        reports_output_dir = tempfile.mkdtemp()
        cls.log_system_intelligence(reports_output_dir)
        cls.log_conda_environment(reports_output_dir)

    @staticmethod
    def md5(fname: str):
        """Generate md5 sum for input file"""
        hash_md5 = hashlib.md5()
        with open(fname, 'rb') as f:
            for chunk in iter(lambda: f.read(4096), b''):
                hash_md5.update(chunk)

        md5sum = hash_md5.hexdigest()
        return md5sum

    @classmethod
    def get_md5_sums(cls, dir, md5_sums=None):
        """
        Recursively go through directories and subdirectories
        and generate tuples of (<file_path>, <md5sum>)
        returns: list of tuples
        """
        if not md5_sums:
            md5_sums = []
        elements = glob.glob(dir + "/*")
        for elem in elements:
            # if file, get md5 sum
            if os.path.isfile(elem):
                elem_md5 = cls.md5(elem)
                md5_sums.append((elem, elem_md5))
            # if directory, apply recursion
            if os.path.isdir(elem):
                md5_sums = cls.get_md5_sums(elem, md5_sums)
            else:
                continue

    @classmethod
    def log_input_data(cls, input_data: str):
        print('[bold blue]Hashing input data...')
        input_hash = cls.get_md5_sums(input_data)
        mlflow.log_param("input_hash", input_data + "-" + input_hash)


sc_autoencoder/data_loading/data_loader.py
import tensorflow as tf
import scanpy as sc


def preprocessing(adata):
    # Perform preprocessing of a anndata object
    sc.pp.filter_cells(adata, min_genes=200)
    sc.pp.filter_genes(adata, min_cells=3)

    # Normalization and scaling:
    sc.pp.normalize_total(adata, target_sum=1e4)
    sc.pp.log1p(adata)

    # Identify highly-variable genes
    sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5, subset=True)
    sc.pp.scale(adata, zero_center=True, max_value=3)
    x = adata.X
    data = tf.data.Dataset.from_tensor_slices((x, x))
    return data, x


def load_data(strategy, batch_size, buffer_size, tensorflow_seed):
    """
    Load a single cell dataset and for feeding to the model
    """
    adata = sc.datasets.pbmc3k()
    adata.var_names_make_unique()

    # preprocessing
    dataset, test_data = preprocessing(adata)

    # Distribute data to devices and scale images
    BATCH_SIZE_PER_REPLICA = batch_size
    BATCH_SIZE = BATCH_SIZE_PER_REPLICA * strategy.num_replicas_in_sync

    dataset = dataset.cache().shuffle(buffer_size, seed=tensorflow_seed, reshuffle_each_iteration=False).batch(BATCH_SIZE)

    return dataset, test_data


sc_autoencoder/visualization/run_experiments.sh
#!/bin/sh
#===============================================================#
# Perform experiments with different reproducibility settings
#===============================================================#

## ID: 1 --> gpu_non_deterministic
#- gpu deterministic settings disabled
echo "Running gpu_non_deterministic"
for var in 1 2 3 4 5 6 7 8 9 10; do mlflow run ../../ -A gpus=all --experiment-name  gpu_non_deterministic -P deterministic='0'; done;

## ID: 2 --> gpu_random
#- gpu deterministic settings disabled
#- setting different seeds every time:
echo "Running gpu_random"
for var in 1 2 3 4 5 6 7 8 9 10; do mlflow run ../../ -A gpus=all -P tensorflow-seed=$var  -P general-seed=$var --experiment-name  gpu_random -P deterministic='0'; done;

## ID: 3 --> cpu_random
#- setting different seeds every time:
echo "Running cpu_random"
for var in 1 2 3 4 5 6 7 8 9 10; do mlflow run ../../ -P tensorflow-seed=$var  -P general-seed=$var --experiment-name  cpu_random; done;

## ID: 4 --> gpu_deterministic
#- normal deterministic settings
echo "Running gpu_deterministic"
for var in 1 2 3 4 5 6 7 8 9 10; do mlflow run ../../ -A gpus=all --experiment-name  gpu_deterministic -P deterministic='1'; done;


sc_autoencoder/model/model.py
import tensorflow as tf


class Autoencoder(tf.keras.Model):
    """
    Autoencoder Model for single-cell RNA-seq data

    :param input_dim:       The dimension of the input data (number of genes)
    :param layer_sizes:     A list of size 3 with the layer sizes of encoder and decoder
    :param latent_size:     The size of the latent (bottleneck) dimension
    """

    def __init__(self, input_dim, layer_sizes, latent_size):
        super(Autoencoder, self).__init__()
        self.input_dim = input_dim
        self.layer_sizes = layer_sizes
        self.latent_size = latent_size

        # Encoder
        self.encoder = tf.keras.Sequential([
            tf.keras.layers.Dense(self.layer_sizes[0],  activation="relu", name="encoder1"),
            tf.keras.layers.Dense(self.layer_sizes[1],  activation="relu", name="encoder2"),
            tf.keras.layers.Dense(self.layer_sizes[2],  activation="relu", name="encoder3"),
            tf.keras.layers.Dense(self.latent_size,     activation="relu", name="latent")
        ])

        # Decoder
        self.decoder = tf.keras.Sequential([
            tf.keras.layers.Dense(self.layer_sizes[2],  activation="relu", name="decoder1"),
            tf.keras.layers.Dense(self.layer_sizes[1],  activation="relu", name="decoder2"),
            tf.keras.layers.Dense(self.layer_sizes[0],  activation="relu", name="decoder3"),
            tf.keras.layers.Dense(self.input_dim,     activation="linear", name="out_layer")
        ])

    def call(self, x):
        encoded = self.encoder(x)
        return self.decoder(encoded)

    def encode(self, x):
        return self.encoder(x)


def create_model(input_shape):
    model = Autoencoder(input_dim=input_shape, layer_sizes=[256, 128, 64], latent_size=32)
    return model
